<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Landscape genetics (classification) • mrIML</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Landscape genetics (classification)">
<meta name="robots" content="noindex">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">mrIML</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">2.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/mrIML.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Landscape%20genetics%20(classification).html">Landscape genetics (classification)</a></li>
    <li><a class="dropdown-item" href="../articles/Regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../articles/Graphical%20network%20model%20(GNM).html">Graphical network model (GNM)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Other</h6></li>
    <li><a class="dropdown-item" href="../articles/Common%20errors%20in%20mrIML.html">Common errors in mrIML</a></li>
    <li><a class="dropdown-item" href="../articles/Microbiome%20data%20curation.html">Microbiome data curation</a></li>
    <li><a class="dropdown-item" href="../articles/Swine%20biosecurity.html">Swine biosecurity</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/nickfountainjones/mrIML/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Landscape genetics (classification)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nickfountainjones/mrIML/blob/master/vignettes/articles/Landscape%20genetics%20(classification).Rmd"><code>vignettes/articles/Landscape genetics (classification).Rmd</code></a></small>
      <div class="d-none name"><code>Landscape genetics (classification).Rmd</code></div>
    </div>

    
    
<p>MrIML is a R package allows users to generate and interpret
multi-response models (i.e., joint species distribution models)
leveraging advances in data science and machine learning. MrIML couples
the tidymodel infrastructure developed by Max Kuhn and colleagues with
model agnostic interpretable machine learning tools to gain insights
into multiple response data such as. As such mrIML is flexible and
easily extendable allowing users to construct everything from simple
linear models to tree-based methods for each response using the same
syntax and same way to compare predictive performance. In this vignette
we will guide you through how to apply this package to landscape
genetics problems but keep in mind any multiple-response data can be
interrogated using this package (e.g., species or OTU presence/absence
data).</p>
<p>Let’s start by loading the required packages and data. The data we
are going to use is from Fountain-Jones etal (2017) and consists of
single nucleotide polymorphism (SNP) data from a virus (feline
immunodeficiency virus, FIV) infecting bobcats in Southern
California.</p>
<p>Features (a.k.a predictor variables) with missing data can be removed
or interpolated depending on the nature of the missing data. Here we
remove them.</p>
<p>When constructing these type of models, SNPs that are common (occur
say in &gt;80% of samples) or rare (&lt; 10% of samples) are difficult
to model. This cn be a problem for potentially adaptice loci which tend
to be rarer We provide down-sampling or up-sampling strategies for
dealing with unbalanced data (see below) but filtering out common and
rare loci is another option. The more response variables (SNPs) you
include the longer the computational time, so for this example we will
trim SNPs occurring in &lt; 40% and &gt; 70% of individuals. This leaves
29 SNPs remaining. In practice, these filtering steps are stringent so
&lt; 20 and &gt; 80 maybe more appropriate.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Optional: Filter rare/common SNPs or species. Retaining minor allelle frequncies &gt;0.1 and removing common allelles (occur&gt;0.9)</span></span>
<span><span class="va">fData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterRareCommon.html">filterRareCommon</a></span><span class="op">(</span><span class="va">Responsedata</span>,</span>
<span>                           lower <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>                           higher <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> </span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">fData</span></span>
<span></span>
<span><span class="co">#for simplicity when comparing</span></span>
<span><span class="co">#another option at this stage is to filter response that are strongly correlated with each other.</span></span>
<span><span class="co">#df2 &lt;- cor(X) #find correlations</span></span>
<span><span class="co">#hc &lt;-  findCorrelation(df2, cutoff=0.5) # put any value as a "cutoff". </span></span>
<span><span class="co">#hc &lt;-  sort(hc)</span></span>
<span><span class="co">#X &lt;-  X[,-c(hc)] #</span></span></code></pre></div>
<p>If your feature data consists of landscape resistance data generated
from Circuitscape or similar, we provide functionality that can extract
each surface from a folder and generate resistance components that can
be used in these models (using principal coordinate analysis). This data
can be easily merged if needs be with other non-matrix data (e.g.,
age/sex). There was no resistance surfaces here, but this is what the
code looks like:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#R &lt;- resist_components(filename = 'location of pairwise matrices', p_val=0.01 ) # p values are used here to filter resistance components that aren't correlated with th original pairwise matrix.</span></span>
<span><span class="co">#X&lt;- cbind(R,X)</span></span></code></pre></div>
<p>Now all the data is loaded and ready to go we can formulate the model
using tidymodel syntax. In this case we have binary data (SNP
presence/absence at each loci) but the data could also be counts or
continuous (the set_model argument would be “regression” instead of
“classification”). The user can specify any model from the tidymodel
universe as ‘model 1’ (see <a href="https://www.tidymodels.org/find/" class="external-link uri">https://www.tidymodels.org/find/</a> for details). However,
we have done most of our testing on random forests (rf), xgr boost and
glms (generalized linear models). Here we will specify a random forest
classification model as the model applied to each response.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_rf</span> <span class="op">&lt;-</span> <span class="fu">rand_forest</span><span class="op">(</span>trees <span class="op">=</span> <span class="fl">100</span>,</span>
<span>              mode <span class="op">=</span> <span class="st">"classification"</span>,</span>
<span>              mtry <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>,</span>
<span>              min_n <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co">#100 trees are set for brevity. Aim to start with 1000</span></span>
<span>             <span class="fu">set_engine</span><span class="op">(</span><span class="st">"randomForest"</span><span class="op">)</span></span></code></pre></div>
<p>Now we can run the model. Hyper parameter tuning (for algorithms that
have hyper parameters) is done automatically by testing how model
performance changes across a random grid of 10 parameters and the best
performing combination is kept. The only choice to make is to either
down/up sample data or leave it as is. As our viral data set is small,
we will not do any up or down-sampling (option ‘no’). We could also
implement parallel processing for larger data sets.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">"multisession"</span>, workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">yhats_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">X</span>,</span>
<span>                          Y <span class="op">=</span> <span class="va">Y</span>,</span>
<span>                          Model <span class="op">=</span> <span class="va">model_rf</span>,</span>
<span>                          tune_grid_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                          k <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                          racing <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>It takes a couple of minutes to run on a laptop and there are
warnings that you can ignore (this data set is small). The above code
constructs tuned random forest models for the features (Y) you have
selected for each response (29 viral SNP in this case). We can then
check the performance of each model separately and overall. The
performance metrics we provide for classification models are area under
curve (roc_AUC), Mathew’s correlation coefficient (MCC - a good metric
when data is imbalanced i.e., unequal numbers of 0/1s in each response),
specificity (proportion correctly classified not having the mutation at
that loci) and sensitivity (proportion correctly classified having the
mutation at that loci). We also provide ‘prevalence’ that tells you how
common that SNP was in the population sampled.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ModelPerf_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_rf</span><span class="op">)</span></span>
<span><span class="va">ModelPerf_rf</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co">#predictive performance for individual responses </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 29 × 8</span></span></span>
<span><span class="co">#&gt;    response model_name roc_AUC    mcc sensitivity     ppv specificity prevalence</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> env_131  rand_fore…   1      0.316        0.25   1           1          0.421</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> env_163  rand_fore…   0.625  0            0    <span style="color: #BB0000;">NaN</span>           1          0.632</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> env_164  rand_fore…   1      0.447        1      0.6         0.333      0.421</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> env_167  rand_fore…   0     -<span style="color: #BB0000;">0.447</span>        0.4    0.667       0          0.421</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> env_169  rand_fore…   1      0.316        1      0.4         0.25       0.421</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> env_212  rand_fore…   0.4   -<span style="color: #BB0000;">0.316</span>        0      0           0.6        0.684</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> env_23   rand_fore…   0.625  0            0    <span style="color: #BB0000;">NaN</span>           1          0.632</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> env_24   rand_fore…   0.75   0            1      0.333       0          0.421</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> env_41   rand_fore…   1      0.5          0.5    1           1          0.474</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> env_47   rand_fore…   1      1            1      1           1          0.474</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 19 more rows</span></span></span>
<span><span class="va">ModelPerf_rf</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="co">#overall predictive performance</span></span>
<span><span class="co">#&gt; [1] 0.1815828</span></span></code></pre></div>
<p>You can compare the predictive performance of random forests with any
other technique simply by altering the following. Lets try a logistic
regression. Note that for logistic regression you may need to
scale/transform the data or create dummy variables if you include
categorical variables. These steps can easily be added to the pipeline
following tidymodel syntax (see the website above)</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_glm</span> <span class="op">&lt;-</span> <span class="co">#model used to generate yhat</span></span>
<span>  <span class="fu">logistic_reg</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">set_engine</span><span class="op">(</span><span class="st">"glm"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">set_mode</span><span class="op">(</span><span class="st">"classification"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">yhats_glm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">X</span>,</span>
<span>                           Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                           Model<span class="op">=</span><span class="va">model_glm</span>,</span>
<span>                           tune_grid_size<span class="op">=</span><span class="fl">5</span>,</span>
<span>                           k<span class="op">=</span><span class="fl">5</span>,</span>
<span>                           racing<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   7%  |                                                                              |=======                                                               |  10%  |                                                                              |==========                                                            |  14%  |                                                                              |============                                                          |  17%  |                                                                              |==============                                                        |  21%  |                                                                              |=================                                                     |  24%  |                                                                              |===================                                                   |  28%  |                                                                              |======================                                                |  31%  |                                                                              |========================                                              |  34%  |                                                                              |===========================                                           |  38%  |                                                                              |=============================                                         |  41%  |                                                                              |===============================                                       |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |=======================================                               |  55%  |                                                                              |=========================================                             |  59%  |                                                                              |===========================================                           |  62%  |                                                                              |==============================================                        |  66%  |                                                                              |================================================                      |  69%  |                                                                              |===================================================                   |  72%  |                                                                              |=====================================================                 |  76%  |                                                                              |========================================================              |  79%  |                                                                              |==========================================================            |  83%  |                                                                              |============================================================          |  86%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  93%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%</span></span>
<span></span>
<span><span class="co">#save(yhats, file='logreg_model')</span></span>
<span><span class="va">ModelPerf_glm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_glm</span><span class="op">)</span> <span class="co">#</span></span>
<span><span class="va">ModelPerf_glm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 0.09562822</span></span></code></pre></div>
<p>You can see that the random forests model outperforms the logistic
regression in this case. Once we are happy with the underlying model we
can then dive in to see how this model predicts the SNPs. First we can
check out variable importance.The first plot is the global importance
(what features shape genetic change overall), the second shows the
individual models (with an optional importance MCC threshold ‘cutoff’
value ). We also provide functionality to assess outlier SNPs utilizing
a PCA on importance scores. We can see only one outlier was detected on
PC3 (env24) indicating that this SNP was bets predicted by a different
set of predictors. Looking at the indiivudal importance plots shows
that, unlike the other SNPs, forest cover was the best predictor for
env24. See the regression vignette for more details.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">glm_impVI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrVip.html">mrVip</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_glm</span>,</span>
<span>  mrBootstrap_obj <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  global_top_var <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  local_top_var <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  taxa <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  model_perf <span class="op">=</span> <span class="va">ModelPerf_glm</span><span class="op">)</span></span>
<span></span>
<span><span class="va">glm_impVI</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="co">#importance</span></span></code></pre></div>
<p><img src="Landscape%20genetics%20(classification)_files/figure-html/unnamed-chunk-8-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">glm_impVI_PCA</span> <span class="op">&lt;-</span> <span class="va">glm_impVI</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/mrVipPCA.html">mrVipPCA</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">glm_impVI_PCA</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co">#PCA plot</span></span></code></pre></div>
<p><img src="Landscape%20genetics%20(classification)_files/figure-html/unnamed-chunk-8-2.png" width="768" style="display: block; margin: auto;">
If your features can be grouped together this can make it easier to
interpret variable importance plots. The order of the groups should
match the order of the columns in the X data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#for example</span></span>
<span><span class="va">groupCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span> <span class="op">(</span><span class="st">"Host_characteristics"</span>, <span class="fl">1</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Urbanisation"</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Vegetation"</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Urbanisation"</span>,<span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Spatial"</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'Host_relatedness'</span>, <span class="fl">6</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span> <span class="op">(</span><span class="st">"Host_characteristics"</span>, <span class="fl">1</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Vegetation"</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Urbanisation"</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>MrIML allows a full suite of model agnostic interpretable machine
learning tools to be applied to individual models or the ‘global’ set.
We utilize the R package flashlight which offers an exciting and ever
growing tool set to interpret these models: <a href="https://cran.r-project.org/web/packages/flashlight/vignettes/flashlight.html" class="external-link uri">https://cran.r-project.org/web/packages/flashlight/vignettes/flashlight.html</a></p>
<p>For example, once we create the flashlight objects, we can plot the
predicted values of a feature for each response.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrFlashlight.html">mrFlashlight</a></span><span class="op">(</span><span class="va">yhats_rf</span>,</span>
<span>                   response <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>                   index<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="co">#index pointing to the SNP of interest (i.e. the first column)</span></span>
<span></span>
<span><span class="va">mfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrFlashlight.html">mrFlashlight</a></span><span class="op">(</span><span class="va">yhats_rf</span>,</span>
<span>                    response <span class="op">=</span> <span class="st">"multi"</span><span class="op">)</span> <span class="co">#the global set</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/flashlight/man/light_scatter.html" class="external-link">light_scatter</a></span><span class="op">(</span><span class="va">fl</span>,</span>
<span>                   v <span class="op">=</span> <span class="st">"Grassland"</span>,</span>
<span>                   type <span class="op">=</span> <span class="st">"predicted"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Landscape%20genetics%20(classification)_files/figure-html/unnamed-chunk-10-1.png" width="768" style="display: block; margin: auto;">
We also wrap some flashlight functionality to visualize the marginal
(i.e. partial dependencies) or conditional (accumulated local effects)
effect of a feature on genetic change. Partial dependencies take longer
to calculate and are more sensitive to correlated features. The first
plot You can see here that they can get different results based on which
plot you use. ALE plots are a better option if your feature set is even
moderately impacted by collinearity ( e.g., rho = 0.6). The second plot
is the overall smoothed genetic-turnover function.</p>
<p>When running this code yourself, you need to change ‘v’ to the
variable of interest within your Y dataset.In this case we are looking
at how grassland shapes genetic turnover.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Partial dependencies</span></span>
<span><span class="va">profilePlot_pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrCovar.html">mrCovar</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_rf</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"Grassland"</span>,</span>
<span>  sdthresh <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span><span class="va">profilePlot_pd</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Landscape%20genetics%20(classification)_files/figure-html/unnamed-chunk-11-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Acumulated local effects</span></span>
<span><span class="va">profilePlot_ale</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrCovar.html">mrCovar</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_rf</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"Grassland"</span>,</span>
<span>  sdthresh <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"ale"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># This second plot is the cumulative turnover function</span></span>
<span><span class="va">profilePlot_ale</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Landscape%20genetics%20(classification)_files/figure-html/unnamed-chunk-11-2.png" width="768" style="display: block; margin: auto;">
Finally, we can assess how features interact overall to shape genetic
change. Be warned this is memory intensive. Future updates to this
package will enable users to visualize these interactions and explore
them in more detail using 2D ALE plots for example.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">interactions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrInteractions.html">mrInteractions</a></span><span class="op">(</span><span class="va">yhats_rf</span>,</span>
<span>                               feature<span class="op">=</span><span class="st">'Forest'</span>,</span>
<span>                               num_bootstrap<span class="op">=</span><span class="fl">1</span>,</span>
<span>                               top_int<span class="op">=</span><span class="fl">2</span><span class="op">)</span> <span class="co">#this is computationally intensive so multicore are needed. </span></span></code></pre></div>
<p>You could easily compare the plots generated above from our random
forest model with the logistic regression model we also calculated.</p>
<p>You can visualize detected interactions by converting the MrIML
predicts object into a form suitable for the R package IML (<a href="https://cran.r-project.org/web/packages/iml/vignettes/intro.html" class="external-link uri">https://cran.r-project.org/web/packages/iml/vignettes/intro.html</a>)
using ‘MrIMLconverts’. You can plot how two features interact to shape
genetic change using ‘mrProfile2D’. In effect, this heatmap shows the
average prediction of the interacting features across loci/response
variables. ALEs or partial dependencies can be used - but we highly
recommend using ALEs for the most robust results (see Molnar (2019) for
more detail)</p>
<p>This is just an example of the tools can be applied to better
interpret multi-response models and in turn gain insights into how
landscape impacts genetic change. We’ve demonstrated that this package
is modular and flexible, allowing future work on ‘tidymodels’ and
interpretable machine learning to easily be incorporated. The
predictions from these models, for example, can be used to identify
individual SNPs that may be adaptive assess gene flow across space
(future updates) or to help better formulate probabilistic models such
as generalized dissimilarity models (GDMs).</p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://discover.utas.edu.au/Nick.FountainJones" class="external-link">Nick Fountain-Jones</a>, Ryan Leadbetter, <a href="https://machado-lab.github.io/" class="external-link">Gustavo Machado</a>, Chris Kozakiewicz, Nick Clark.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
