<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Multi-response (regression) • mrIML</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multi-response (regression)">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">mrIML</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">2.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/mrIML.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Multi-response%20(classification).html">Multi-response (classification)</a></li>
    <li><a class="dropdown-item" href="../articles/Multi-response%20(regression).html">Multi-response (regression)</a></li>
    <li><a class="dropdown-item" href="../articles/Graphical%20network%20model%20(GNM).html">Graphical network model (GNM)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Other</h6></li>
    <li><a class="dropdown-item" href="../articles/Common%20errors%20in%20mrIML.html">Common errors in mrIML</a></li>
    <li><a class="dropdown-item" href="../articles/Microbiome%20data%20curation.html">Microbiome data curation</a></li>
    <li><a class="dropdown-item" href="../articles/Swine%20biosecurity.html">Swine biosecurity</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/nickfountainjones/mrIML/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Multi-response (regression)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nickfountainjones/mrIML/blob/master/vignettes/articles/Multi-response%20(regression).Rmd"><code>vignettes/articles/Multi-response (regression).Rmd</code></a></small>
      <div class="d-none name"><code>Multi-response (regression).Rmd</code></div>
    </div>

    
    
<p><code>mrIML</code> is an R package that allows users to generate and
interpret multi-response models (i.e., joint species distribution
models) leveraging advances in data science and machine learning.
<code>mrIML</code> couples the <a href="https://www.tidymodels.org" class="external-link"><code>tidymodels</code></a>
infrastructure, developed by Max Kuhn and colleagues, with
model-agnostic interpretable machine learning tools to gain insights
into multiple response data. <code>mrIML</code> is flexible and easily
extendable, allowing users to construct everything from simple linear
models to tree-based methods for each response using the same syntax and
compare them under the same predictive performance criteria.</p>
<p>In this vignette, we will guide you through how to apply this package
to ecological genomics problems using the regression functionality of
the package. The data set we’ll use comes from Fitzpatrick et al. 2014,
who examined adaptive genetic variation in relation to geography and
climate adaptation (current and future) in balsam poplar (Populus
balsamifera). See Ecology Letters, (2014) doi: 10.1111/ele.12376. In
this paper, they used the similar gradient forests routine (see Ellis et
al. 2012 Ecology), and we show that <code>mrIML</code> can not only
provide more flexible model choice and interpretive capabilities but can
derive new insights into the relationship between climate and genetic
variation. Further, we show that linear models of each loci have
slightly greater predictive performance.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in data file with minor allele freqs &amp; env/space variables</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"gfData.RData"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get climate &amp; MEM variables for predictors</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="va">gfData</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"bio_"</span><span class="op">)</span>,</span>
<span>    <span class="st">"elevation"</span>,</span>
<span>    <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"MEM."</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_if</a></span><span class="op">(</span><span class="va">is.integer</span>, <span class="va">as.numeric</span><span class="op">)</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="va">gfData</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"GI5"</span><span class="op">)</span><span class="op">)</span> <span class="co"># GIGANTEA-5 (GI5)</span></span></code></pre></div>
<p>We focus on the adaptive SNP loci from GIGANTEA-5 (GI5) gene that has
known links to stem development, plant circadian clock, and light
perception pathway. The data represents the proportion of individuals in
each population with that SNP loci.</p>
<div class="section level3">
<h3 id="parallel-processing">Parallel processing<a class="anchor" aria-label="anchor" href="#parallel-processing"></a>
</h3>
<p><code>mrIML</code> uses the flexible <a href="https://future.apply.futureverse.org" class="external-link"><code>future.apply::future_lapply()</code></a>
to set up multi-core processing. This can greatly speed up the more
computationally expensive <code>mrIML</code> functions. In the example
below, we set up a cluster using 4 cores. If you don’t set up a cluster,
the default settings will be used and the analysis will run
sequentially.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">"multisession"</span>, workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="building-and-comparing-models">Building and comparing models<a class="anchor" aria-label="anchor" href="#building-and-comparing-models"></a>
</h3>
<p>Performing the analysis is very similar to our classification
example. Let’s start by constructing a linear model for this data set.
We set Model 1 to a linear regression. See <a href="https://www.tidymodels.org/find/" class="external-link uri">https://www.tidymodels.org/find/</a> for other regression
model options. Note that ‘mode’ must be ‘regression’ and in
<code>mrIML</code>predicts, model has to be set to ‘regression’.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_lm</span> <span class="op">&lt;-</span> <span class="fu">linear_reg</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">set_engine</span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">set_mode</span><span class="op">(</span><span class="st">"regression"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">yhats_lm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span></span>
<span>  X <span class="op">=</span> <span class="va">X</span>,</span>
<span>  Y <span class="op">=</span> <span class="va">Y</span>,</span>
<span>  Model <span class="op">=</span> <span class="va">model_lm</span>,</span>
<span>  prop <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>  tune_grid_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  racing <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Model performance can be examined the same way as in the <a href="https://nickfountainjones.github.io/mrIML/articles/Multi-response%20(classification).html" class="external-link">classification
example</a>, however the reported performance metrics are different.
Running <code><a href="../reference/mrIMLperformance.html">mrIMLperformance()</a></code> on a regression model provides
the root mean square error (rmse) and R².</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ModelPerf_lm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_lm</span><span class="op">)</span></span>
<span><span class="va">ModelPerf_lm</span><span class="op">$</span><span class="va">model_performance</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 4</span></span></span>
<span><span class="co">#&gt;    response           model_name   rmse  rsquared</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> CANDIDATE_GI5_108  linear_reg 0.044<span style="text-decoration: underline;">0</span>  0.290   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> CANDIDATE_GI5_198  linear_reg 0.140   0.392   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> CANDIDATE_GI5_268  linear_reg 0.084<span style="text-decoration: underline;">7</span>  0.182   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> CANDIDATE_GI5_92   linear_reg 0.096<span style="text-decoration: underline;">5</span>  0.116   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> CANDIDATE_GI5_1950 linear_reg 0.243   0.000<span style="text-decoration: underline;">340</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> CANDIDATE_GI5_2382 linear_reg 0.137   0.008<span style="text-decoration: underline;">96</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> CANDIDATE_GI5_2405 linear_reg 0.145   0.398   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> CANDIDATE_GI5_2612 linear_reg 0.120   0.585   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> CANDIDATE_GI5_2641 linear_reg 0.066<span style="text-decoration: underline;">9</span>  0.120   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> CANDIDATE_GI5_33   linear_reg 0.167   0.618   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> CANDIDATE_GI5_3966 linear_reg 0.178   0.024<span style="text-decoration: underline;">4</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> CANDIDATE_GI5_5033 linear_reg 0.072<span style="text-decoration: underline;">5</span>  0.002<span style="text-decoration: underline;">77</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> CANDIDATE_GI5_5090 linear_reg 0.136   0.262   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> CANDIDATE_GI5_5119 linear_reg 0.121   0.011<span style="text-decoration: underline;">8</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> CANDIDATE_GI5_8997 linear_reg 0.194   0.233   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> CANDIDATE_GI5_9287 linear_reg 0.063<span style="text-decoration: underline;">2</span>  0.015<span style="text-decoration: underline;">0</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">17</span> CANDIDATE_GI5_9447 linear_reg 0.081<span style="text-decoration: underline;">6</span> <span style="color: #BB0000;">NA</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">18</span> CANDIDATE_GI5_9551 linear_reg 0.151   0.192   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">19</span> CANDIDATE_GI5_9585 linear_reg 0.177   0.321   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">20</span> CANDIDATE_GI5_9659 linear_reg 0.151   0.637</span></span>
<span><span class="va">ModelPerf_lm</span><span class="op">$</span><span class="va">global_performance_summary</span></span>
<span><span class="co">#&gt; [1] 0.1284043</span></span></code></pre></div>
<p>You can see that the overall R² is 0.13 but there is substantial
variation in predictive performance across loci.</p>
<p>Let’s compare the performance of the linear model to that of a random
forest. Random forest is the computational engine in gradient forests.
Notice for random forests we have two hyperparameters to tune:
<code>mtry</code> (number of features to randomly include at each split)
and <code>min_n</code> (the minimum number of data points in a node that
are required for the node to be split further). The syntax <span style="color:red"><code>tune()</code></span> acts as a placeholder to
tell <code>mrIML</code> to tune those hyperparameters across a grid of
values (defined in <code>mrIMLpredicts</code>
<code>tune_grid_size</code> argument). Different algorithms will have
different hyperparameters; see <a href="https://www.tidymodels.org/find/parsnip/" class="external-link uri">https://www.tidymodels.org/find/parsnip/</a> for parameter
details. Note that large grid sizes (&gt;10) for algorithms with lots of
hyperparameters (such as extreme gradient boosting) will be
computationally demanding. In this case we choose a grid size of 5.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_rf</span> <span class="op">&lt;-</span> <span class="fu">rand_forest</span><span class="op">(</span></span>
<span>  trees <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"regression"</span>,</span>
<span>  mtry <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  min_n <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span> <span class="fu">set_engine</span><span class="op">(</span><span class="st">"randomForest"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">yhats_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span></span>
<span>  X <span class="op">=</span> <span class="va">X</span>,</span>
<span>  Y <span class="op">=</span> <span class="va">Y</span>,</span>
<span>  Model <span class="op">=</span> <span class="va">model_rf</span>,</span>
<span>  tune_grid_size <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   5%  |                                                                              |=======                                                               |  10%  |                                                                              |==========                                                            |  15%  |                                                                              |==============                                                        |  20%  |                                                                              |==================                                                    |  25%  |                                                                              |=====================                                                 |  30%  |                                                                              |========================                                              |  35%  |                                                                              |============================                                          |  40%  |                                                                              |================================                                      |  45%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================                                |  55%  |                                                                              |==========================================                            |  60%  |                                                                              |==============================================                        |  65%  |                                                                              |=================================================                     |  70%  |                                                                              |====================================================                  |  75%  |                                                                              |========================================================              |  80%  |                                                                              |============================================================          |  85%  |                                                                              |===============================================================       |  90%  |                                                                              |==================================================================    |  95%  |                                                                              |======================================================================| 100%</span></span>
<span></span>
<span><span class="va">ModelPerf_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_rf</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ModelPerf_rf</span><span class="op">$</span><span class="va">model_performance</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 4</span></span></span>
<span><span class="co">#&gt;    response           model_name    rmse rsquared</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> CANDIDATE_GI5_108  rand_forest 0.030<span style="text-decoration: underline;">9</span>  0.073<span style="text-decoration: underline;">8</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> CANDIDATE_GI5_198  rand_forest 0.080<span style="text-decoration: underline;">4</span>  0.508  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> CANDIDATE_GI5_268  rand_forest 0.075<span style="text-decoration: underline;">7</span>  0.658  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> CANDIDATE_GI5_92   rand_forest 0.160   0.352  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> CANDIDATE_GI5_1950 rand_forest 0.141   0.222  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> CANDIDATE_GI5_2382 rand_forest 0.065<span style="text-decoration: underline;">0</span>  0.014<span style="text-decoration: underline;">8</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> CANDIDATE_GI5_2405 rand_forest 0.109   0.556  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> CANDIDATE_GI5_2612 rand_forest 0.145   0.080<span style="text-decoration: underline;">9</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> CANDIDATE_GI5_2641 rand_forest 0.082<span style="text-decoration: underline;">5</span>  0.002<span style="text-decoration: underline;">93</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> CANDIDATE_GI5_33   rand_forest 0.154   0.285  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> CANDIDATE_GI5_3966 rand_forest 0.116   0.206  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> CANDIDATE_GI5_5033 rand_forest 0.043<span style="text-decoration: underline;">8</span>  0.418  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> CANDIDATE_GI5_5090 rand_forest 0.101   0.474  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> CANDIDATE_GI5_5119 rand_forest 0.124   0.093<span style="text-decoration: underline;">3</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> CANDIDATE_GI5_8997 rand_forest 0.168   0.449  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> CANDIDATE_GI5_9287 rand_forest 0.052<span style="text-decoration: underline;">8</span>  0.051<span style="text-decoration: underline;">7</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">17</span> CANDIDATE_GI5_9447 rand_forest 0.144   0.007<span style="text-decoration: underline;">88</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">18</span> CANDIDATE_GI5_9551 rand_forest 0.116   0.456  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">19</span> CANDIDATE_GI5_9585 rand_forest 0.140   0.235  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">20</span> CANDIDATE_GI5_9659 rand_forest 0.126   0.371</span></span>
<span><span class="va">ModelPerf_rf</span><span class="op">$</span><span class="va">global_performance_summary</span></span>
<span><span class="co">#&gt; [1] 0.1086905</span></span>
<span></span>
<span><span class="co">#easier to see with plots</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrPerformancePlot.html">mrPerformancePlot</a></span><span class="op">(</span></span>
<span>  ModelPerf1 <span class="op">=</span> <span class="va">ModelPerf_lm</span>,</span>
<span>  ModelPerf2 <span class="op">=</span> <span class="va">ModelPerf_rf</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"regression"</span></span>
<span><span class="op">)</span> </span>
<span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">/</span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Multi-response%20(regression)_files/figure-html/fit-rf-1.png" width="768" style="display: block; margin: auto;"></p>
<p>You can see that predictive performance is actually slightly less
using RF (overall R² = 0.11) but for some loci RF does better than our
LM and sometimes worse. Which to choose? Generally, the simpler the
better, the linear model in this case, but it depends on how important
you think non-linear responses are. In future versions of
<code>mrIML</code> we will implement ensemble models that will overcome
this issue. However, for the time being, we’ll need to interrogate the
model and make our own informed decision.</p>
</div>
<div class="section level3">
<h3 id="interpreting-your-models">Interpreting your models<a class="anchor" aria-label="anchor" href="#interpreting-your-models"></a>
</h3>
<p>A first step is to look at variable importance. We do this below for
the RF model using <code><a href="../reference/mrVip.html">mrVip()</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrVip.html">mrVip</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_rf</span>,</span>
<span>  mrBootstrap_obj <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  global_top_var <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  local_top_var <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  taxa <span class="op">=</span> <span class="st">"CANDIDATE_GI5_9585"</span></span>
<span><span class="op">)</span> </span>
<span></span>
<span><span class="va">VI</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Multi-response%20(regression)_files/figure-html/rf-vip-1.png" width="960" style="display: block; margin: auto;"></p>
<p>In the above figure, the top left plot shows the ranked most
important variables across all models collectively, while the subplots
in the top right show the most important variables in the top few models
(the number of variables and models to visualize can be controlled using
the <code>global_top_var</code>, <code>local_top_var</code>, and
<code>threshold</code> arguments of <code><a href="../reference/mrVip.html">mrVip()</a></code>, see
<code>?mrVip()</code>). The bottom plot shows the most important
variables in a specific response model, which we defined by
<code>taxa = "CANDIDATE_GI5_9585"</code>. Generally, <code>bio_10</code>
(mean summer temperature), <code>bio_1</code> (mean annual temperature),
and <code>bio_18</code> (summer precipitation) are the most influential
predictor variables, however this varies quite a lot across the loci.
Summer precipitation was not as important in Fitzpatrick et al. but
otherwise these results are similar.</p>
<p>We can also perform PCA of the variable importance scores in the
different models to group loci that behave similarly and to identify
outliers.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># PCA</span></span>
<span><span class="va">VI_PCA</span> <span class="op">&lt;-</span> <span class="va">VI</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/mrVipPCA.html">mrVipPCA</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">VI_PCA</span><span class="op">$</span><span class="va">eigenvalues</span></span>
<span><span class="co">#&gt;  [1] 3.35524390 2.35918680 1.80331596 0.97732841 0.71118104 0.64171481</span></span>
<span><span class="co">#&gt;  [7] 0.49911497 0.26273404 0.19623715 0.14511906 0.04882385</span></span>
<span><span class="va">VI_PCA</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; Warning: ggrepel: 4 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">#&gt; increasing max.overlaps</span></span></code></pre></div>
<p><img src="Multi-response%20(regression)_files/figure-html/rf-vip-PCA-1.png" width="768" style="display: block; margin: auto;"></p>
<p>The right plot of the first two PCs shows that candidate 5119, 9287,
5033, 92, and 108 are shaped similarly by the features we included and
may, for example, be products of linked selection. The left plot shows
that most of the variability in the variable importance data is captured
by the first three principal components.</p>
<p>Note that you can also supply bootstraps for importance scores in
<code><a href="../reference/mrVip.html">mrVip()</a></code>, <span style="color:red">but this functionality is
still under development for regression models.</span></p>
<p>Next, we can explore the model further by plotting the relationships
between our SNPs and a feature in our set. Let’s choose
<code>bio_1</code> (mean annual temperature) and plot the individual and
global (average of all SNPs) partial dependency (PD) plots.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PD_bio1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrCovar.html">mrCovar</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_rf</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"bio_1"</span>,</span>
<span>  sdthresh <span class="op">=</span> <span class="fl">0.01</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="op">(</span><span class="va">PD_bio1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.4</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="op">(</span><span class="va">PD_bio1</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.4</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="op">(</span><span class="va">PD_bio1</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">y</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">y</span>, which will replace the existing scale.</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">y</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">y</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="Multi-response%20(regression)_files/figure-html/bio_1-PD-1.png" width="768" style="display: block; margin: auto;"></p>
<p>The top plot in the above figure shows partial dependency curves for
SNPs that respond to mean annual temperature. What we mean by “respond”
here is that the prediction surface (the line) deviates across the Y
axis of the PD plots. We measure this deviation by calculating the
standard deviation and use that as a threshold
(<code>sdthresh = 0.01</code> in this case, and this will differ by data
set). The middle plot is the average partial dependency curve of all
SNPs across an annual temperature gradient. The individual PDs are shown
as grey silhouettes in the background. This is very similar to the
pattern observed by Fitzpatrick et al., <span style="color:red">except
with a slight decline in SNP turnover with mean annual temperatures &gt;
0.</span> Combined, you can see here only a few candidate SNPs are
driving this pattern and these may warrant further interrogation.</p>
<p>The bottom plot in the figure shows the rate of change over the SNP
PD curves. There is a general rise from -80 to -20. This kind of plot
can help to identify general thresholds.</p>
<p>Let’s compare the PDs to accumulated local effect (ALE) plots that
are less sensitive to correlations among features (see Molnar 2019). We
generate ALE plots by supplying <code>type = "ale"</code> to
<code><a href="../reference/mrCovar.html">mrCovar()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ALE_bio1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrCovar.html">mrCovar</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_rf</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"bio_1"</span>,</span>
<span>  sdthresh <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"ale"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="op">(</span><span class="va">ALE_bio1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.4</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="op">(</span><span class="va">ALE_bio1</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.4</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span></span>
<span>  <span class="op">(</span><span class="va">ALE_bio1</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">y</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">y</span>, which will replace the existing scale.</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">y</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">y</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="Multi-response%20(regression)_files/figure-html/bio_1-ale-1.png" width="768" style="display: block; margin: auto;"></p>
<p>The effect of mean annual temperature on SNP turnover is not as
distinct in the global ALE plot. This may mean that correlations between
features may be important for the predictions.</p>
</div>
<div class="section level3">
<h3 id="other-interpretation-methods">Other interpretation methods<a class="anchor" aria-label="anchor" href="#other-interpretation-methods"></a>
</h3>
<p><span style="color:red"><code>mrIML</code> has easy-to-use
functionality that can quantify interactions between features. Note that
this can take a while to compute and will be the topic of future
work.</span></p>
<p>This is touching only the surface of what is possible in terms of
interrogating this model. Both Flashlight and IML packages have a wide
variety of tools that can offer novel insights into how these models
perform. See <a href="https://cran.r-project.org/web/packages/flashlight/vignettes/flashlight.html" class="external-link uri">https://cran.r-project.org/web/packages/flashlight/vignettes/flashlight.html</a>
and <a href="https://cran.r-project.org/web/packages/iml/vignettes/intro.html" class="external-link uri">https://cran.r-project.org/web/packages/iml/vignettes/intro.html</a>
for other options.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://discover.utas.edu.au/Nick.FountainJones" class="external-link">Nick Fountain-Jones</a>, Ryan Leadbetter, <a href="https://machado-lab.github.io/" class="external-link">Gustavo Machado</a>, Chris Kozakiewicz, Nick Clark.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
