<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Multi-response (classification) • mrIML</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multi-response (classification)">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">mrIML</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">2.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/mrIML.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Multi-response%20(classification).html">Multi-response (classification)</a></li>
    <li><a class="dropdown-item" href="../articles/Multi-response%20(regression).html">Multi-response (regression)</a></li>
    <li><a class="dropdown-item" href="../articles/Graphical%20network%20model%20(GNM).html">Graphical network model (GNM)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Other</h6></li>
    <li><a class="dropdown-item" href="../articles/Common%20errors%20in%20mrIML.html">Common errors in mrIML</a></li>
    <li><a class="dropdown-item" href="../articles/Microbiome%20data%20curation.html">Microbiome data curation</a></li>
    <li><a class="dropdown-item" href="../articles/Swine%20biosecurity.html">Swine biosecurity</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/nickfountainjones/mrIML/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Multi-response (classification)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nickfountainjones/mrIML/blob/master/vignettes/articles/Multi-response%20(classification).Rmd"><code>vignettes/articles/Multi-response (classification).Rmd</code></a></small>
      <div class="d-none name"><code>Multi-response (classification).Rmd</code></div>
    </div>

    
    
<p><code>mrIML</code> is an R package that allows users to generate and
interpret multi-response models (i.e., joint species distribution
models) leveraging advances in data science and machine learning.
<code>mrIML</code> couples the <a href="https://www.tidymodels.org" class="external-link"><code>tidymodels</code></a>
infrastructure developed by Max Kuhn and colleagues with model-agnostic
interpretable machine learning tools to gain insights into multiple
response data. As such, <code>mrIML</code> is flexible and easily
extendable, allowing users to construct everything from simple linear
models to tree-based methods for each response using the same syntax and
comparisons of predictive performance. In this vignette, we will guide
you through how to apply this package to landscape genetics problems,
but keep in mind that any multiple-response data can be interrogated
using this package (e.g., species or OTU presence/absence data).</p>
<p>Let’s start by loading the required packages and data. The data we
are going to use is from Fountain-Jones et al. (2017) and consists of
single nucleotide polymorphism (SNP) data from a virus (feline
immunodeficiency virus, FIV) infecting bobcats in Southern
California.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/nickfountainjones/mrIML">mrIML</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidymodels.tidymodels.org" class="external-link">tidymodels</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Attaching packages</span> ────────────────────────────────────── tidymodels 1.3.0 ──</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">broom       </span> 1.0.8     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">recipes     </span> 1.3.0</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">dials       </span> 1.4.0     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">rsample     </span> 1.3.0</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">dplyr       </span> 1.1.4     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tibble      </span> 3.2.1</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">ggplot2     </span> 3.5.2     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tidyr       </span> 1.3.1</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">infer       </span> 1.0.8     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">tune        </span> 1.3.0</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">modeldata   </span> 1.4.0     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">workflows   </span> 1.2.0</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">parsnip     </span> 1.3.1     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">workflowsets</span> 1.1.0</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">purrr       </span> 1.0.4     <span style="color: #00BB00;">✔</span> <span style="color: #0000BB;">yardstick   </span> 1.3.2</span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Conflicts</span> ───────────────────────────────────────── tidymodels_conflicts() ──</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">purrr</span>::<span style="color: #00BB00;">discard()</span> masks <span style="color: #0000BB;">scales</span>::discard()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">filter()</span>  masks <span style="color: #0000BB;">stats</span>::filter()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">dplyr</span>::<span style="color: #00BB00;">lag()</span>     masks <span style="color: #0000BB;">stats</span>::lag()</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="color: #0000BB;">recipes</span>::<span style="color: #00BB00;">step()</span>  masks <span style="color: #0000BB;">stats</span>::step()</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mayer79/flashlight" class="external-link">flashlight</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">7007</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"Features.RData"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"Responsedata.RData"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Print first 6 cols</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Features</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="co">#&gt;      Grassland Shrub.Scrub      Forest   HighlyDev       Urban    Suburban</span></span>
<span><span class="co">#&gt; X21 0.07001059   0.5569084 0.010719958 0.000000000 0.000000000 0.003573319</span></span>
<span><span class="co">#&gt; X22 0.06766419   0.7673464 0.030587924 0.000000000 0.000000000 0.132680085</span></span>
<span><span class="co">#&gt; X23 0.18449304   0.5236581 0.008614977 0.002253148 0.001590457 0.013253810</span></span>
<span><span class="co">#&gt; X24 0.09806776   0.7858655 0.000661726 0.000000000 0.000000000 0.001191106</span></span>
<span><span class="co">#&gt; X27 0.15776500   0.8416187 0.000616270 0.000000000 0.000000000 0.000000000</span></span>
<span><span class="co">#&gt; X31 0.08730321   0.6391331 0.008791658 0.000000000 0.000000000 0.007360458</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Responsedata</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="co">#&gt;     env_1 env_10 env_100 env_104 env_107 env_108</span></span>
<span><span class="co">#&gt; X21     0      0       1       0       0       0</span></span>
<span><span class="co">#&gt; X22     0      0       0       0       0       0</span></span>
<span><span class="co">#&gt; X23     0      0       0       0       1       0</span></span>
<span><span class="co">#&gt; X24     0      0       0       1       0       0</span></span>
<span><span class="co">#&gt; X27     0      0       0       0       1       0</span></span>
<span><span class="co">#&gt; X31     1      0       0       1       0       0</span></span></code></pre></div>
<p>When constructing these types of models, SNPs that are common (occur
in &gt;80% of samples) or rare (&lt;10% of samples) are difficult to
model. This can be a problem for potentially adaptive loci, which tend
to be rarer. We provide down-sampling or up-sampling strategies for
dealing with unbalanced data (see below), but filtering out common and
rare loci is another option. The more response variables (SNPs) you
include, the longer the computational time, so for this example, we will
drop SNPs occurring in &lt;40% or &gt;70% of individuals. This leaves 29
SNPs remaining. In practice, these filtering steps are stringent, so
&lt;20% and &gt;80% may be more appropriate.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define the responses</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterRareCommon.html">filterRareCommon</a></span><span class="op">(</span></span>
<span>  <span class="va">Responsedata</span>,</span>
<span>  lower <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  higher <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Define the predictors</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="va">Features</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html" class="external-link">where</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html" class="external-link">n_distinct</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>If your feature data consists of landscape resistance data generated
from Circuitscape or similar, we provide functionality that can extract
each surface from a folder and generate resistance components that can
be used in these models (using principal coordinate analysis). This data
can be easily merged with other non-matrix data (e.g., age/sex) if
needed. There were no resistance surfaces here, but if there were, we
could load them as follows:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">R</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/resist_components.html">resist_components</a></span><span class="op">(</span></span>
<span>  filename <span class="op">=</span> <span class="st">"FILE_PATH"</span>, p_val <span class="op">=</span> <span class="fl">0.01</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add to predictors</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">R</span>, <span class="va">X</span><span class="op">)</span></span></code></pre></div>
<p>Now all the data is loaded and ready to go we can formulate the model
using tidymodel syntax. In this case we have binary data (SNP
presence/absence at each loci) but the data could also be counts or
continuous (the <code>mode</code> argument would be
<code>"regression"</code> instead of <code>"classification"</code>). The
user can specify any model from the tidymodel universe (see <a href="https://www.tidymodels.org/find/" class="external-link uri">https://www.tidymodels.org/find/</a> for details). However,
we have done most of our testing on random forests (RF), xgr boost and
glms (generalized linear models). Here we will specify an RF
classification model to fit each response.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_rf</span> <span class="op">&lt;-</span> <span class="fu">rand_forest</span><span class="op">(</span></span>
<span>  trees <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"classification"</span>,</span>
<span>  mtry <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  min_n <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span> <span class="fu">set_engine</span><span class="op">(</span><span class="st">"randomForest"</span><span class="op">)</span></span></code></pre></div>
<p>Now we can run the model. Hyper parameter tuning (for algorithms that
have hyper parameters) is done automatically by testing how model
performance changes across a random grid and the best performing
combination is kept. The only choice to make is to either down/up sample
data or leave it as is. As our viral data set is small, we will not do
any up- or down-sampling (i.e.,<code>balance_data = "no"</code>). We can
also implement parallel processing for larger data sets to speed up the
<code>mrIML</code> workflow.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">"multisession"</span>, workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">yhats_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span></span>
<span>  X <span class="op">=</span> <span class="va">X</span>,</span>
<span>  Y <span class="op">=</span> <span class="va">Y</span>,</span>
<span>  Model <span class="op">=</span> <span class="va">model_rf</span>,</span>
<span>  tune_grid_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  racing <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>It takes a couple of minutes to run on a laptop and there are
warnings that you can ignore (this data set is small). The above code
constructs tuned random forest models for the features (<code>Y</code>)
you have selected for each response (29 viral SNP in this case). We can
then check the performance of each model separately and overall. The
performance metrics we provide for classification models are area under
curve (<code>roc_AUC</code>), Mathew’s correlation coefficient
(<code>mcc</code> - a good metric when data is imbalanced i.e., unequal
numbers of 0/1s in each response), specificity (proportion correctly
classified not having the mutation at that loci) and sensitivity
(proportion correctly classified having the mutation at that loci). We
also provide “prevalence”, which tells you how common that SNP was in
the population sampled.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ModelPerf_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_rf</span><span class="op">)</span></span>
<span><span class="va">ModelPerf_rf</span><span class="op">$</span><span class="va">model_performance</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 29 × 8</span></span></span>
<span><span class="co">#&gt;    response model_name roc_AUC    mcc sensitivity     ppv specificity prevalence</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> env_131  rand_fore…   0.625  0.25         0.75   0.75          0.5      0.421</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> env_163  rand_fore…   1      0            0    <span style="color: #BB0000;">NaN</span>             1        0.632</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> env_164  rand_fore…   0.444  0            1      0.5           0        0.421</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> env_167  rand_fore…   0     -<span style="color: #BB0000;">0.632</span>        0.2    0.5           0        0.421</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> env_169  rand_fore…   0.875  0            1      0.333         0        0.421</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> env_212  rand_fore…   0      0            0    <span style="color: #BB0000;">NaN</span>             1        0.684</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> env_23   rand_fore…   0.5    0.316        0.25   1             1        0.632</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> env_24   rand_fore…   0.5    0            1      0.333         0        0.421</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> env_41   rand_fore…   0.625  0            0.5    0.667         0.5      0.474</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> env_47   rand_fore…   1      1            1      1             1        0.474</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 19 more rows</span></span></span>
<span><span class="va">ModelPerf_rf</span><span class="op">$</span><span class="va">global_performance_summary</span></span>
<span><span class="co">#&gt; [1] 0.1128229</span></span></code></pre></div>
<p>You can compare the predictive performance of random forests with any
other technique simply by altering the following. Lets try a logistic
regression. Note that for logistic regression you may need to
scale/transform the data or create dummy variables if you include
categorical variables. These steps can easily be added to the pipeline
following tidymodel syntax.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define LM</span></span>
<span><span class="va">model_glm</span> <span class="op">&lt;-</span> <span class="fu">logistic_reg</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">set_engine</span><span class="op">(</span><span class="st">"glm"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit mrIML model</span></span>
<span><span class="va">yhats_glm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span></span>
<span>  X <span class="op">=</span> <span class="va">X</span>,</span>
<span>  Y <span class="op">=</span> <span class="va">Y</span>,</span>
<span>  Model <span class="op">=</span> <span class="va">model_glm</span>,</span>
<span>  tune_grid_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  racing <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |==                                                                    |   3%  |                                                                              |=====                                                                 |   7%  |                                                                              |=======                                                               |  10%  |                                                                              |==========                                                            |  14%  |                                                                              |============                                                          |  17%  |                                                                              |==============                                                        |  21%  |                                                                              |=================                                                     |  24%  |                                                                              |===================                                                   |  28%  |                                                                              |======================                                                |  31%  |                                                                              |========================                                              |  34%  |                                                                              |===========================                                           |  38%  |                                                                              |=============================                                         |  41%  |                                                                              |===============================                                       |  45%  |                                                                              |==================================                                    |  48%  |                                                                              |====================================                                  |  52%  |                                                                              |=======================================                               |  55%  |                                                                              |=========================================                             |  59%  |                                                                              |===========================================                           |  62%  |                                                                              |==============================================                        |  66%  |                                                                              |================================================                      |  69%  |                                                                              |===================================================                   |  72%  |                                                                              |=====================================================                 |  76%  |                                                                              |========================================================              |  79%  |                                                                              |==========================================================            |  83%  |                                                                              |============================================================          |  86%  |                                                                              |===============================================================       |  90%  |                                                                              |=================================================================     |  93%  |                                                                              |====================================================================  |  97%  |                                                                              |======================================================================| 100%</span></span>
<span></span>
<span><span class="co"># Get performance</span></span>
<span><span class="va">ModelPerf_glm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_glm</span><span class="op">)</span></span>
<span><span class="va">ModelPerf_glm</span><span class="op">$</span><span class="va">global_performance_summary</span></span>
<span><span class="co">#&gt; [1] 0.1907751</span></span></code></pre></div>
<p>You can see that the random forests model outperforms the logistic
regression in this case (0.1907751 versus 0.1128229). Once we are happy
with the underlying model we can then dive in to see how this model
predicts the SNPs. First we can check variable importance.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">glm_impVI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrVip.html">mrVip</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_glm</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  global_top_var <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  local_top_var <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  model_perf <span class="op">=</span> <span class="va">ModelPerf_glm</span></span>
<span><span class="op">)</span></span>
<span><span class="va">glm_impVI</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Multi-response%20(classification)_files/figure-html/var-imp-1.png" width="768" style="display: block; margin: auto;"></p>
<p>The output of <code><a href="../reference/mrVip.html">mrVip()</a></code> contains a figure summarising
variable imporance. The main plot on the left shows the global
importance (what features shape genetic change overall) and the sub
plots on the right show the individual models (with an option to filter
by the mcc value using <code>threshold</code>). We also provide
functionality to assess outlier SNPs utilizing a PCA on importance
scores.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">glm_impVI_PCA</span> <span class="op">&lt;-</span> <span class="va">glm_impVI</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/mrVipPCA.html">mrVipPCA</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">glm_impVI_PCA</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; Warning: ggrepel: 1 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">#&gt; increasing max.overlaps</span></span></code></pre></div>
<p><img src="Multi-response%20(classification)_files/figure-html/unnamed-chunk-2-1.png" width="768" style="display: block; margin: auto;"></p>
<p><span style="color:red">We can see only one outlier was detected on
PC3 (env24) indicating that this SNP was bets predicted by a different
set of predictors. Looking at the indiivudal importance plots shows
that, unlike the other SNPs, forest cover was the best predictor for
env24. See the regression vignette for more details.</span></p>
<p>If your features can be grouped together this can make it easier to
interpret variable importance plots. The order of the groups should
match the order of the columns in the X data, for example:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">groupCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Host_characteristics"</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Urbanisation"</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Vegetation"</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Urbanisation"</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Spatial"</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Host_relatedness"</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Host_characteristics"</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Vegetation"</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Urbanisation"</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code>mrIML</code> allows a full suite of model agnostic
interpretable machine learning tools to be applied to individual models
or the “global” throught <code><a href="../reference/mrFlashlight.html">mrFlashlight()</a></code>. We utilize the R
package <a href="https://mayer79.github.io/flashlight/" class="external-link"><code>flashlight</code></a>
which offers an exciting and ever growing tool set to interpret these
kinds of models. For example, once we create the flashlight objects, we
can plot the predicted values of a feature for each response.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrFlashlight.html">mrFlashlight</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_rf</span>,</span>
<span>  response <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>  index <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">fl</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/flashlight/man/light_scatter.html" class="external-link">light_scatter</a></span><span class="op">(</span></span>
<span>    v <span class="op">=</span> <span class="st">"Grassland"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"predicted"</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Multi-response%20(classification)_files/figure-html/flashlight-1.png" width="768" style="display: block; margin: auto;"></p>
<p><code>mrIMl</code> also provides a wraper function for some
flashlight functionality to visualize the marginal (i.e. partial
dependencies) or conditional (accumulated local effects) effect of a
feature on the responses. Partial dependencies take longer to calculate
and are more sensitive to correlated features, therefore the results of
the two can vary. For example, in the two plots bellow generated with
<code><a href="../reference/mrCovar.html">mrCovar()</a></code>, the results differ slightly but not
significantly. ALE plots are a better option if your feature set is even
moderately impacted by collinearity (e.g., rho = 0.6). <span style="color:red">The second plot is the overall smoothed
genetic-turnover function.</span></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Partial dependencies</span></span>
<span><span class="va">profilePlot_pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrCovar.html">mrCovar</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_rf</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"Grassland"</span>,</span>
<span>  sdthresh <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span><span class="va">profilePlot_pd</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<p><img src="Multi-response%20(classification)_files/figure-html/marginal-effects-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Acumulated local effects</span></span>
<span><span class="va">profilePlot_ale</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrCovar.html">mrCovar</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_rf</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"Grassland"</span>,</span>
<span>  sdthresh <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"ale"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">profilePlot_ale</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<p><img src="Multi-response%20(classification)_files/figure-html/marginal-effects-2.png" width="768" style="display: block; margin: auto;"></p>
<p>Finally, we can assess how features interact overall to shape genetic
change using <code><a href="../reference/mrInteractions.html">mrInteractions()</a></code>. Be warned, this is memory
intensive. Future updates to this package will enable users to visualize
these interactions and explore them in more detail using 2D ALE plots
for example.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">interactions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrInteractions.html">mrInteractions</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_rf</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"Forest"</span>,</span>
<span>  num_bootstrap <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  top_int <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">interactions</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">/</span></span>
<span>  <span class="va">interactions</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">/</span></span>
<span>  <span class="va">interactions</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Multi-response%20(classification)_files/figure-html/interactions-1.png" width="768" style="display: block; margin: auto;"></p>
<p>You could easily compare the plots generated above from our random
forest model with the logistic regression model we fit earlier.</p>
<p>You can visualize detected interactions by converting the
<code>mrIML</code> object into a flashlight object and using the
<code><a href="https://rdrr.io/pkg/flashlight/man/light_profile2d.html" class="external-link">flashlight::light_profile2d()</a></code> function. In effect, the
resulting heatmap shows the average prediction of the interacting
features across loci/response variables. Both ALEs and partial
dependencies can be used - but we highly recommend using ALEs for the
most robust results (see Molnar (2019) for more detail)</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrFlashlight.html">mrFlashlight</a></span><span class="op">(</span></span>
<span>  <span class="va">yhats_rf</span>,</span>
<span>  response <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>  index <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">fl</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/flashlight/man/light_profile2d.html" class="external-link">light_profile2d</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Forest"</span>, <span class="st">"Grassland"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">Y</span>, <span class="va">X</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Multi-response%20(classification)_files/figure-html/two-way-interaction-1.png" width="768" style="display: block; margin: auto;"></p>
<p>This is just an example of the tools can be applied to better
interpret multi-response models and in turn gain insights into how
landscape impacts genetic change. We’ve demonstrated that this package
is modular and flexible, allowing future work on ‘tidymodels’ and
interpretable machine learning to easily be incorporated. The
predictions from these models, for example, can be used to identify
individual SNPs that may be adaptive assess gene flow across space
(future updates) or to help better formulate probabilistic models such
as generalized dissimilarity models (GDMs).</p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://discover.utas.edu.au/Nick.FountainJones" class="external-link">Nick Fountain-Jones</a>, Ryan Leadbetter, <a href="https://machado-lab.github.io/" class="external-link">Gustavo Machado</a>, Chris Kozakiewicz, Nick Clark.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
