<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Regression • mrIML</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Regression">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">mrIML</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/mrIML.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Landscape%20genetics%20(classification).html">Landscape genetics (classification)</a></li>
    <li><a class="dropdown-item" href="../articles/Regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../articles/Graphical%20network%20model%20(GNM).html">Graphical network model (GNM)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Other</h6></li>
    <li><a class="dropdown-item" href="../articles/Common%20errors%20in%20mrIML.html">Common errors in mrIML</a></li>
    <li><a class="dropdown-item" href="../articles/Microbiome%20data%20curation.html">Microbiome data curation</a></li>
    <li><a class="dropdown-item" href="../articles/Swine%20biosecurity.html">Swine biosecurity</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/nickfountainjones/mrIML/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Regression</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nickfountainjones/mrIML/blob/master/vignettes/articles/Regression.Rmd"><code>vignettes/articles/Regression.Rmd</code></a></small>
      <div class="d-none name"><code>Regression.Rmd</code></div>
    </div>

    
    
<p>MrIML is a R package allows users to generate and interpret
multi-response models (i.e., joint species distribution models)
leveraging advances in data science and machine learning. MrIML couples
the tidymodel infrastructure developed by Max Kuhn and colleagues with
model agnostic interpretable machine learning tools to gain insights
into multiple response data such as. As such mrIML is flexible and
easily extendable allowing users to construct everything from simple
linear models to tree-based methods for each response using the same
syntax and same way to compare predictive performance. In this vignette
we will guide you through how to apply this package to ecological
genomics problems using the regression functionality of the package.
This data set comes from Fitzpatrick et al 2014 who were examining
adaptive genetic variation in relation to geography and climate
adaptation (current and future) in balsam poplar(Populus balsamifera).
See Ecology Letters, (2014) doi: 10.1111/ele.12376. In this paper they
used the similar gradient forests routine (see Ellis et al 2012
Ecology), and we show that MrIML can not only provide more flexible
model choice and interpretive capabilities, but can derive new insights
into the relationship between climate and genetic variation. Further, we
show that linear models of each loci have slightly greater predictive
performance.</p>
<p>We focus on the adaptive SNP loci from GIGANTEA-5 (GI5) gene that has
known links to stem development, plant circadian clock and light
perception pathway. The data is the proportion of individuals in that
population with that SNP loci.</p>
<div class="section level3">
<h3 id="lets-load-that-data">Lets load that data<a class="anchor" aria-label="anchor" href="#lets-load-that-data"></a>
</h3>
<p>###Parallel processing</p>
<p>MrIML provides uses the flexible future apply functionality to set up
multi-core processing. In the example below, we set up a cluster using 4
cores. If you don’t set up a cluster, the default settings will be used
and the analysis will run sequentially.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span>, workers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="running-the-analyis">Running the analyis<a class="anchor" aria-label="anchor" href="#running-the-analyis"></a>
</h3>
<p>Performing the analysis is very similar to our classification
example. Lets start with a constructing a linear model for this data
set. We set Model 1 to a linear regression. See <a href="https://www.tidymodels.org/find/" class="external-link uri">https://www.tidymodels.org/find/</a> for other regression
model options Note that ‘mode’ must be regression and in MrIMLpredicts,
model has to be set to ‘regression’.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_lm</span> <span class="op">&lt;-</span> <span class="co">#model used to generate yhat</span></span>
<span>  <span class="co"># specify that the model is a random forest</span></span>
<span>  <span class="fu">linear_reg</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># select the engine/package that underlies the model</span></span>
<span>  <span class="fu">set_engine</span><span class="op">(</span><span class="st">"lm"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># choose either the continuous regression or binary classification mode</span></span>
<span>  <span class="fu">set_mode</span><span class="op">(</span><span class="st">"regression"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">yhats_lm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">X</span>,</span>
<span>                          Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                          X1<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                          Model<span class="op">=</span><span class="va">model_lm</span>,</span>
<span>                          prop<span class="op">=</span><span class="fl">0.7</span>,</span>
<span>                          tune_grid_size<span class="op">=</span> <span class="fl">10</span>,</span>
<span>                          k<span class="op">=</span><span class="fl">10</span>,</span>
<span>                          racing<span class="op">=</span><span class="cn">F</span><span class="op">)</span> <span class="co">## Balanced data= up updamples and down downsampled to create a balanced set. For regression 'no' has to be selected.</span></span>
<span><span class="co">#racing can't be used for linear models - no parameters to tune</span></span></code></pre></div>
<p>Model performance can be examined the same way as in the
classification example, however the metrics are different. We provide
root mean square error (rmse) and R2. You can see that the overall R2 is
0.13 but there is substantial variation across loci in predictive
performance.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ModelPerf_lm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_lm</span><span class="op">)</span></span>
<span><span class="va">ModelPerf_lm</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co">#predictive performance for individual responses. </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 4</span></span></span>
<span><span class="co">#&gt;    response           model_name   rmse rsquared</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> CANDIDATE_GI5_108  linear_reg 0.030<span style="text-decoration: underline;">7</span> 0.187   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> CANDIDATE_GI5_198  linear_reg 0.133  0.331   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> CANDIDATE_GI5_268  linear_reg 0.090<span style="text-decoration: underline;">7</span> 0.354   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> CANDIDATE_GI5_92   linear_reg 0.138  0.273   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> CANDIDATE_GI5_1950 linear_reg 0.193  0.307   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> CANDIDATE_GI5_2382 linear_reg 0.105  0.023<span style="text-decoration: underline;">6</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> CANDIDATE_GI5_2405 linear_reg 0.133  0.431   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> CANDIDATE_GI5_2612 linear_reg 0.181  0.726   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> CANDIDATE_GI5_2641 linear_reg 0.079<span style="text-decoration: underline;">8</span> 0.046<span style="text-decoration: underline;">0</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> CANDIDATE_GI5_33   linear_reg 0.155  0.409   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> CANDIDATE_GI5_3966 linear_reg 0.195  0.000<span style="text-decoration: underline;">540</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> CANDIDATE_GI5_5033 linear_reg 0.073<span style="text-decoration: underline;">0</span> 0.002<span style="text-decoration: underline;">08</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> CANDIDATE_GI5_5090 linear_reg 0.167  0.415   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> CANDIDATE_GI5_5119 linear_reg 0.056<span style="text-decoration: underline;">7</span> 0.121   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> CANDIDATE_GI5_8997 linear_reg 0.114  0.566   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> CANDIDATE_GI5_9287 linear_reg 0.058<span style="text-decoration: underline;">0</span> 0.002<span style="text-decoration: underline;">56</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">17</span> CANDIDATE_GI5_9447 linear_reg 0.097<span style="text-decoration: underline;">0</span> 0.429   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">18</span> CANDIDATE_GI5_9551 linear_reg 0.164  0.680   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">19</span> CANDIDATE_GI5_9585 linear_reg 0.142  0.254   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">20</span> CANDIDATE_GI5_9659 linear_reg 0.153  0.000<span style="text-decoration: underline;">684</span></span></span>
<span><span class="va">ModelPerf_lm</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="co">#overall average r2 </span></span>
<span><span class="co">#&gt; [1] 0.1228556</span></span></code></pre></div>
<p>Lets compare the performance of linear models to that of random
forests. Random forests is the computational engine in gradient forests.
Notice for random forests we have two hyperparameters to tune; mtry
(number of features to randomly include at each split) and min_n (the
minimum number of data points in a node that are required for the node
to be split further). The syntax ‘tune()’ acts a placeholder to tell
MrIML to tune those hyperparamters across a grid of values (defined in
MRIML predicts ‘tune_grid_size’ argument). Different algorithms will
have different hyperparameters.See <a href="https://www.tidymodels.org/find/parsnip/" class="external-link uri">https://www.tidymodels.org/find/parsnip/</a> for parameter
details. Note that large grid sizes (&gt;10) for algorithms with lots of
hyperparameters (such as extreme gradient boosting) will be
computationally demanding. In this case we choose a grid size of 5.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_rf</span> <span class="op">&lt;-</span> <span class="fu">rand_forest</span><span class="op">(</span>trees <span class="op">=</span> <span class="fl">100</span>,</span>
<span>              mode <span class="op">=</span> <span class="st">"regression"</span>,</span>
<span>              mtry <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span>,</span>
<span>              min_n <span class="op">=</span> <span class="fu">tune</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co">#100 trees are set for brevity. Aim to start with 1000</span></span>
<span>         <span class="fu">set_engine</span><span class="op">(</span><span class="st">"randomForest"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">yhats_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLpredicts.html">mrIMLpredicts</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">X</span>,</span>
<span>                          Y<span class="op">=</span><span class="va">Y</span>,</span>
<span>                          Model<span class="op">=</span><span class="va">model_rf</span>,</span>
<span>                          tune_grid_size<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |====                                                                  |   5%  |                                                                              |=======                                                               |  10%  |                                                                              |==========                                                            |  15%  |                                                                              |==============                                                        |  20%  |                                                                              |==================                                                    |  25%  |                                                                              |=====================                                                 |  30%  |                                                                              |========================                                              |  35%  |                                                                              |============================                                          |  40%  |                                                                              |================================                                      |  45%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================                                |  55%  |                                                                              |==========================================                            |  60%  |                                                                              |==============================================                        |  65%  |                                                                              |=================================================                     |  70%  |                                                                              |====================================================                  |  75%  |                                                                              |========================================================              |  80%  |                                                                              |============================================================          |  85%  |                                                                              |===============================================================       |  90%  |                                                                              |==================================================================    |  95%  |                                                                              |======================================================================| 100%</span></span>
<span></span>
<span><span class="va">ModelPerf_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrIMLperformance.html">mrIMLperformance</a></span><span class="op">(</span><span class="va">yhats_rf</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ModelPerf_rf</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co">#predictive performance for individual responses. </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 20 × 4</span></span></span>
<span><span class="co">#&gt;    response           model_name    rmse rsquared</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> CANDIDATE_GI5_108  rand_forest 0.024<span style="text-decoration: underline;">8</span>  0.072<span style="text-decoration: underline;">0</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> CANDIDATE_GI5_198  rand_forest 0.144   0.366  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> CANDIDATE_GI5_268  rand_forest 0.083<span style="text-decoration: underline;">8</span>  0.083<span style="text-decoration: underline;">9</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> CANDIDATE_GI5_92   rand_forest 0.050<span style="text-decoration: underline;">2</span>  0.451  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> CANDIDATE_GI5_1950 rand_forest 0.190   0.437  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> CANDIDATE_GI5_2382 rand_forest 0.068<span style="text-decoration: underline;">9</span>  0.317  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> CANDIDATE_GI5_2405 rand_forest 0.172   0.308  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> CANDIDATE_GI5_2612 rand_forest 0.142   0.260  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> CANDIDATE_GI5_2641 rand_forest 0.077<span style="text-decoration: underline;">3</span>  0.007<span style="text-decoration: underline;">12</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> CANDIDATE_GI5_33   rand_forest 0.131   0.014<span style="text-decoration: underline;">0</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> CANDIDATE_GI5_3966 rand_forest 0.150   0.049<span style="text-decoration: underline;">3</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> CANDIDATE_GI5_5033 rand_forest 0.029<span style="text-decoration: underline;">0</span>  0.010<span style="text-decoration: underline;">2</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> CANDIDATE_GI5_5090 rand_forest 0.152   0.733  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> CANDIDATE_GI5_5119 rand_forest 0.114   0.151  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> CANDIDATE_GI5_8997 rand_forest 0.105   0.657  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> CANDIDATE_GI5_9287 rand_forest 0.039<span style="text-decoration: underline;">0</span>  0.066<span style="text-decoration: underline;">0</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">17</span> CANDIDATE_GI5_9447 rand_forest 0.080<span style="text-decoration: underline;">6</span>  0.080<span style="text-decoration: underline;">5</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">18</span> CANDIDATE_GI5_9551 rand_forest 0.113   0.123  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">19</span> CANDIDATE_GI5_9585 rand_forest 0.138   0.340  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">20</span> CANDIDATE_GI5_9659 rand_forest 0.118   0.295</span></span>
<span><span class="va">ModelPerf_rf</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="co">#overall average r2 </span></span>
<span><span class="co">#&gt; [1] 0.1061354</span></span>
<span></span>
<span><span class="co">#easier to see with plots</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrPerformancePlot.html">mrPerformancePlot</a></span><span class="op">(</span>ModelPerf1<span class="op">=</span><span class="va">ModelPerf_lm</span>,</span>
<span>                           ModelPerf2 <span class="op">=</span> <span class="va">ModelPerf_rf</span>,</span>
<span>                           mode<span class="op">=</span><span class="st">'regression'</span> <span class="op">)</span> </span>
<span></span>
<span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Regression_files/figure-html/unnamed-chunk-4-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Regression_files/figure-html/unnamed-chunk-4-2.png" width="768" style="display: block; margin: auto;">
## Plotting</p>
<p>You can see that predictive performance is actually slightly less
using random forests (overall R2 = 0.12) but for some loci random
forests does better than our linear models and sometimes worse. Which to
choose? Generally simpler models are preferred (the linear model in this
case) but it depends on how important to think non-linear response are.
In future versions of MrIML we will implement ensemble models that will
overcome this issue. For the time-being we will have a look at variable
importance for the random forest based model.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">VI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrVip.html">mrVip</a></span><span class="op">(</span><span class="va">yhats_rf</span>,</span>
<span>            mrBootstrap_obj <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>            threshold <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>            global_top_var <span class="op">=</span> <span class="fl">10</span>,</span>
<span>            local_top_var <span class="op">=</span> <span class="fl">5</span>,</span>
<span>            taxa<span class="op">=</span><span class="st">'CANDIDATE_GI5_9585'</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">VI</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="co">#Importance plot</span></span></code></pre></div>
<p><img src="Regression_files/figure-html/unnamed-chunk-5-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">VI_PCA</span> <span class="op">&lt;-</span> <span class="va">VI</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/mrVipPCA.html">mrVipPCA</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">VI_PCA</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Regression_files/figure-html/unnamed-chunk-5-2.png" width="768" style="display: block; margin: auto;">
Cutoff reduces the number of individual SNP plots presented in the
second plot and ‘plot.pca=’yes’’ enables the variable importance scores
to be analysed using principal component analysis (PCA) where SNPs
closer in PCA space are shaped by similar combinations of features. You
can see that bio_18 (summer precipitation), bio_1 (mean annual
temperature) and bio_10 (mean summer temperature) are the most important
features overall. Summer precipitation was not as important in
Fitzpatrick et al but otherwise these results are similar. The second
plot shows the individual models (with an r2 &gt; 0.1, for your data you
will need to play around with this threshold) and you can see for some
SNPs bio_1 is more important whereas for another MEM.1 is more
prominent.The PCA shows that candidate 5119, 9287, 5033 and 108 are
shaped similarly by the features we included and may, for example, be
product of linked selection.</p>
<p>Note that you can also calculate bootstraps for importance scores
(and partial dependencies), but this functionality is still under
development for regression models.</p>
<p>Now we can explore the model further my plotting the relationships
between our SNPs and a feature in our set. Lets choose bio_1 (mean
annual temperature) and plot the individual and global (average of all
SNPs) partial dependency (PD) plots.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PD_bio1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrCovar.html">mrCovar</a></span><span class="op">(</span><span class="va">yhats_rf</span>,</span>
<span>        var <span class="op">=</span> <span class="st">"bio_1"</span>,</span>
<span>        sdthresh <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="va">PD_bio1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Regression_files/figure-html/unnamed-chunk-6-1.png" width="768" style="display: block; margin: auto;">
The first plot is a partial dependency for all SNPs that respond to mean
annual temperature. What we mean by respond here is that the prediction
surface (the line) deviates across the Y axis of the PD plots. We
measure this deviation by calculating the standard deviation and use
that as a threshold (‘sd thresh=0.01’ in this case and this will differ
by data set) to ease visualization of these relationships. The second
plot is the smoothed average partial dependency of SNPs across a annual
temperature gradient. This is very similar to the pattern observed by
Fitzpatrick et al except with a slight decline in SNP turnover with mean
annual temperatures &gt; 0. Combined,you can see here only few candidate
SNPs are driving this pattern and these may warrant further
interrogation.</p>
<p>Lets compare the PDs to accumulated local effect plots that are less
sensitive to correlations among features (see Molnar 2019).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PD_bio_1_ale</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrCovar.html">mrCovar</a></span><span class="op">(</span><span class="va">yhats_rf</span>,</span>
<span>        var <span class="op">=</span> <span class="st">"bio_1"</span>,</span>
<span>        sdthresh <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>        type <span class="op">=</span> <span class="st">"ale"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">PD_bio_1_ale</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="Regression_files/figure-html/unnamed-chunk-7-1.png" width="768" style="display: block; margin: auto;">
The effect of mean annual temperature on SNP turnover is not as distinct
in the global ALE plot. This may mean that correlations between features
may be important for the predictions.</p>
<p>MrIML has easy to use functionality that can can quantify
interactions between features. Note that this can take a while to
compute and will be the topic of a future work.</p>
<p>This is touching only the surface of what is possible in terms of
interrogating this model. Both Flashlight and IML packages have a wide
variety of tools that can offer novel insights into how these models
perform. See <a href="https://cran.r-project.org/web/packages/flashlight/vignettes/flashlight.html" class="external-link uri">https://cran.r-project.org/web/packages/flashlight/vignettes/flashlight.html</a>
and <a href="https://cran.r-project.org/web/packages/iml/vignettes/intro.html" class="external-link uri">https://cran.r-project.org/web/packages/iml/vignettes/intro.html</a>
for other options.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://discover.utas.edu.au/Nick.FountainJones" class="external-link">Nick Fountain-Jones</a>, <a href="https://machado-lab.github.io/" class="external-link">Gustavo Machado</a>, Chris Kozakiewicz, Nick Clark.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
